### PR TITLE
Prevent non repeatable annotations from being propagated more than once

### PR DESCRIPTION
**Pull request by thmasker**
I've encountered this issue with Lombok when using recently upgraded version when compiling files: `lombok.NonNull is not a repeatable annotation type`. After digging a bit, I've found out that some annotations (such as Lombok's `NonNull`. I suspect it happens on all annotations which use `@Target` for both type and variable `ElementType`s) are copied twice because `AbstractBuilderGenerator#setterParameterSpec` calls twice `AbstractBuilderGenerator#addAnnotationsToParam` (second time for type annotations). This throws an error if those annotations are not repeatable. Thus, adding a check to ignore already copied annotations solves this issue

### CODE REVIEW

**Comment by skinny85** [comment]
Can you please add a test, showing what the problem was?

**Comment by thmasker** [comment]
I will try but I think I can't since `ElementType.TYPE_USE` was introduced in Java 8... it will make tests to fail

**Comment by skinny85** [comment]
It won't. The tests run under Java 8+.

**Comment by thmasker** [comment]
I've added a test. I've basically copied Lombok's `@NonNull` annotation, and check that it is only propagated once. Obviously the error does not happen here. If you want to reproduce it, simply try to compile a project using `@NonRepeatableAnnotation` or Lombok's `@NonNull` and a compile error will be thrown

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 1]
```diff
@@ -338,15 +340,17 @@ protected final ParameterSpec setterParameterSpec(VariableElement attribute, Typ
...
         return ret.build();
     }
 
-    private void addAnnotationsToParam(ParameterSpec.Builder param, List<? extends AnnotationMirror> annotations) {
+    private void addAnnotationsToParam(ParameterSpec.Builder param, List<? extends AnnotationMirror> annotations, Set<String> presentAnnotations) {
         for (AnnotationMirror annotation : annotations) {
-            if (this.isAnnotationAllowedOnParam(annotation)) {
+            if (!presentAnnotations.contains(annotation.toString()) && this.isAnnotationAllowedOnParam(annotation)) {
                 param.addAnnotation(AnnotationSpec.get(annotation));
+                presentAnnotations.add(annotation.toString());
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:353]: What if the annotation is repeatable? In that case, should we repeat it?

**Review by skinny85 — COMMENTED** [review]
Thanks for the fix @thmasker! This is a complicated issue, and I think we'll need to iterate on the proper fix a few times before we settle on the correct solution.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java (hunk 1 above)
Comment (thmasker) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java]: I've added a test to check this. It worked exactly the same as non-repeatable annotations, so they would be copied twice. I don't think this test is compatible, so I'll remove it if you want me to

Comment (thmasker) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java]: I don't really know why, if tests run on 8+...

File: build.gradle [hunk 2]
```diff
@@ -87,7 +87,7 @@ repositories {
 if (JavaVersion.current() <= JavaVersion.VERSION_11) {
```
Comment (thmasker) [inline build.gradle:87]: just out of curiosity: why did you put `VERSION_11` instead of `VERSION_1_8`? Optionals were introduced in 1.8, right?

**Comment by skinny85** [comment]
Hey @thmasker,

I've pushed an alternative solution ([here](https://github.com/skinny85/jilt/commit/3105b43d1be0a52896dcba365d4ebdc9a1f2bf1e)), and released it as version `1.6.1.rc1`.

Would you mind testing whether the project you originally ran into this issue with gets fixed by updating to `1.6.1.rc1`?

**Comment by thmasker** [comment]
Just tested it @skinny85 and it works!

**Comment by skinny85** [comment]
OK, this has been fixed in `1.6.1` that was just published.

Thanks for reporting @thmasker, and submitting a PR with the fix!

**Comment by thmasker** [comment]
Thank you @skinny85 for this library! I'm absolutely loving it

**Comment by thmasker** [comment]
I guess it takes some time for Maven central to update, but latest version appearing for now is 1.6.1.rc1, dated Jun 20th...

**Comment by skinny85** [comment]
Yeah, it usually takes a while for the search indexes to be updated. You can see it here: https://search.maven.org/artifact/cc.jilt/jilt/1.6.1/jar

**Pull request closed without merging**